runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/DeadlockDetector.java

```diff
@@ -304,23 +307,31 @@ void lockAcquired(Thread owner, ISchedulingRule lock) {
 		 * or conflict with a lock the given lock will acquire implicitly
 		 * (locks are acquired implicitly when a conflicting lock is acquired)
 		 */
-		ArrayList<ISchedulingRule> conflicting = new ArrayList<>(1);
-		//only need two passes through all the locks to pick up all conflicting rules
-		int NUM_PASSES = 2;
-		conflicting.add(lock);
 		graph[threadIndex][lockIndex]++;
-		for (int i = 0; i < NUM_PASSES; i++) {
-			for (int k = 0; k < conflicting.size(); k++) {
-				ISchedulingRule current = conflicting.get(k);
-				for (int j = 0; j < locks.size(); j++) {
-					ISchedulingRule possible = locks.get(j);
-					if (!conflicting.contains(possible) && current.isConflicting(possible)) {
-						conflicting.add(possible);
-						graph[threadIndex][j]++;
-					}
+
+		// first pass tests against lock:
+		Collection<ISchedulingRule> conflicting = Set.of(lock);
+		conflicting = computeConflicting(threadIndex, conflicting);
+
+		// second pass tests also transitive:
+		if (!conflicting.isEmpty()) {
+			conflicting.add(lock);
+			conflicting = computeConflicting(threadIndex, conflicting);
```

Comment:
```suggestion
			computeConflicting(threadIndex, conflicting);
```

Reply:
Yes one could do that (and i started like that), but then reading the code does not make visible that the same pattern is used for multiple passes.